### ISSUE TITLE
partitioner operation

### ISSUE THREAD

**Issue by omerb01** [pywren/pywren-ibm-cloud#89]
when I try to split a text file from COS into partitions, I think PyWren divides the bytes ranges wrongly. I figured out on file `partitioner.py` the following code in `line 103`:
```python
if chunk_size is not None and obj_size > chunk_size:
    size = 0
    while size < obj_size:
        brange = (size, size+chunk_size+CHUNK_THRESHOLD)
        size += chunk_size
        partition = {}
        partition['map_func_args'] = entry
        partition['data_byte_range'] = brange
        partitions.append(partition)
        total_partitions = total_partitions + 1
```
so I understand that in addition to the `chunk_size`, another variable is added to the range: `CHUNK_THRESHOLD`. is it necessary? or maybe the `CHUNK_THRESHOLD` should be added to `size` variable?
because of this addition, I get some duplications of text in different chunks.

**Comment by gilv** [comment]
@JosepSampe CHUNK_THRESHOLD suppose to be less then row length , right?

**Comment by JosepSampe** [comment]
At this moment, `partitioner` module only makes partitions based on `chunk_size + CHUNK_THRESHOLD`. `CHUNK_THRESHOLD` is used to ensure that there is no split row at the end of each part. `Partitioner` doesn't know the content of the data, so it cannot make partitions based on the content. At the same time, `partitioner` doesn't know the length of the row (different rows may have different sizes), so it is for this reason that  `CHUNK_THRESHOLD` is fixed to a certain size.

To summarize, at this moment the user is the responsible to discard the data at the end of each part, because the beginning of the next part will contain that data.

**Comment by gilv** [comment]
@JosepSampe may be we need partioner to work like this:
 read chunk + threashold long enough. Partioner internally search for the first "new line" in the chunk. When it found one, the rest is ignored.

Any invocation start reading in the middle of the line, ignore partial line. 

This way it's alligned with both scenarios

How is this sounds?

**Comment by JosepSampe** [comment]
@gilv This means that partitioner have to request data to COS, right? Then, partitoner internally search for the first "new line" in the added threshold (to ignore the rest of the threshold) not in the whole chunk, right?

**Comment by gilv** [comment]
@JosepSampe yes...partioner will need to read some small range from COS, to know where line starts and ends..right, but partioner doesn't need to read entire chunk. 
I don't sure actually...doing this way we add more requests, but it's more clean. Doing separation logic inside user's code, less requests, but very messy

**Comment by JosepSampe** [comment]
@gilv Just one thing.. Some datasets contains fields with "new lines" inside. For example, the Airbnb dataset contains a filed called "Comment" where the users may write comments of 1000 characters with "new lines" inside them. Then, how to ensure that the found "new line" is the end of the row, and not a "new line" added as part of the content of a field?

**Comment by gilv** [comment]
@JosepSampe just interesting thing ;) hmm..... I guess if  "new line separator" inside some brackets, then it's not really a new line.. what do you think?

**Comment by JosepSampe** [comment]
@gilv I reviewed the airbnb dataset and comments are enclosed between double quotes: `"comments..."`. So yes, we can follow this approach.

**Comment by gilv** [comment]
@JosepSampe so to summarize, the downside of this approach is that partitioner will likely to make 2 additional requests to the object storage, althoigh it will read very small sizes only to identify data 'borders', but if only counting REST calls - it will be about 2 more REST calls per chunck... correct?

**Comment by JosepSampe** [comment]
For the moment, in #91 I avoid this approach to query 2 times-per-partition to COS. 
Now it can discard cut lines and unnecessary CHUNK_THRESHOLD on-the-fly.

**Comment by JosepSampe** [comment]
@omerb01 I think this can be close since next step is #16
